Data ops

chStaiger opened the pull request:
Implements:

- Up and Download of file/dataobject
- Up and Download of directory/collection
- Determine size of an iRODS collection
- Some useful path-like functions to work with data objects ad collections

maartenschermer commented:
Looks good! I've got a few remarks:

- Analogous to the `iget` and `iput` commands, maybe `irods_get()` and `download_collection()` can be folded into one `iget()` (or `get()`) function, that first figures out whether its an object or a collection and then handles it as required? Same for `irods_put()` and `upload_collection()`, into one `iput()` (or `put()`).
- Downloading a collection without overwrite halts with an exception on the first existing file it encounters. Maybe that's a little harsh, and would it be better to just skip an existing file (with a message) and move on to the next entry. That way, you avoid half downloaded folders.
- Downloading collection downloads the *contents* of the collection, but not the collection (folder) itself, whereas `iget -r <collection>` does create a local folder `<collection>` and downloads the contents into that. Maybe we should replicate that behaviour.
- `irods_put()` seems to overwrite by default, maybe we should add a `overwrite`-flag there as well.

maartenschermer commented on ibridges/irodsconnector/data_operations.py:
```diff
@@ -61,13 +59,13 @@ def obj_replicas(obj: irods.data_object.iRODSDataObject) -> list(tuple(int, str,
         replicas = []
         for r in obj.replicas:
-            replicas.append((r.number, r.resource_name, r.checksum))
+            replicas.append((r.number, r.resource_name, r.checksum, r.size, r.status))
```
Is there a way to check for the user what the value of `status` means (I just get a `1`)?

chStaiger replied:
It is hard to retrieve that from the irods server software. This is how far we got:
OBJ_STATUS_SYMBOL = {
    '0': 'X', # replica checksum check failure
    '1': '&', # replica checksums in sync with other replicas
}
I also found some higher codes. For interpretation I would say, if the status is 1 all is fine, everything else is fishy.

maartenschermer replied:
This is from the `ils` help:
```
   0. X: stale        - The replica is no longer known to be good. Usually [...]
   1. &: good         - The replica is good. Usually the latest replica to be [...]
   2. ?: intermediate - The replica is actively being written to. The state of [...]
   3. ?: write-locked - One of this replica's sibling replicas is actively [...]
```
Maybe we can add the labels (`stale', `good` etc) for clarity (and users don't mistake 0 for the "best" status).

maartenschermer commented on ibridges/irodsconnector/data_operations.py:
```diff
@@ -85,23 +83,35 @@ def get_collection(self, path: IrodsPath) -> irods.collection.iRODSCollection:
 
-    def irods_put(self, local_path: str, irods_path: str, resc_name: str = ''):
+    @staticmethod
+    def has_type_dataobj(item) -> bool:
```
I think `is_dataobj()` might be a more intuitive name. (idem `has_type_collection()` --> `is_collection()`)

chStaiger replied:
I used these function for the IrodsPath verification if a path points to a collections or object. Any suggestions?

chStaiger replied:
I now renamed the type functions as you suggested and renamed the IrodsPath functions to dataobject_exists/collection_exists. This should do the trick.

maartenschermer commented on ibridges/irodsconnector/data_operations.py:
```diff
@@ -110,56 +120,157 @@ def irods_put(self, local_path: str, irods_path: str, resc_name: str = ''):
+        coll = self.get_collection(irods_path)
+        all_objs = self._get_data_objects(coll)
+
+        for o in all_objs:
```
Changing `for o in all_objs:` to something like `folder, file, _, _ in  all_objs:` might be a little more clear.

chStaiger replied:
Thanks!!! Adjusted in both walk loops (directories and collections)

maartenschermer commented on ibridges/irodsconnector/data_operations.py:
```diff
@@ -110,56 +120,157 @@ def irods_put(self, local_path: str, irods_path: str, resc_name: str = ''):
+            Total size [bytes] of the iRODS object or all iRODS objects in the collection.
+
+        """
+        if self.has_type_dataobj(item):
```
Because of the return-statements, you don't really need the `elif` (just `if`), and `else`.

chStaiger replied:
Done. Just in case someone wants more information about the exception, we can add it again later.

maartenschermer commented on ibridges/irodsconnector/data_operations.py:
```diff
@@ -110,56 +120,157 @@ def irods_put(self, local_path: str, irods_path: str, resc_name: str = ''):
+
+        return objs
+
+    def create_collection(self, coll_path: Union[IrodsPath, str]) -> irods.collection.iRODSCollection:
```
`create_collection()` conveniently accepts both `str` and `IrodsPath`, maybe something similar can be implemented for the `get`- and `put`-functions?

chStaiger replied:
I adjusted that to all functions

chStaiger commented:
> Looks good! I've got a few remarks:
> 
>     * Analogous to the `iget` and `iput` commands, maybe `irods_get()` and `download_collection()` can be folded into one `iget()` (or `get()`) function, that first figures out whether its an object or a collection and then handles it as required? Same for `irods_put()` and `upload_collection()`, into one `iput()` (or `put()`).

Since we also have a get_dataobject and get_ collection I named the functions upload and download and the irods_put and irods_get became _obj_put and _obj_get respectively. These two functions we will also need when we want to implement the synchronisation. Let me know if the naming is logically.

> 
>     * Downloading a collection without overwrite halts with an exception on the first existing file it encounters. Maybe that's a little harsh, and would it be better to just skip an existing file (with a message) and move on to the next entry. That way, you avoid half downloaded folders.

Good point! Fixed :)

> 
>     * Downloading collection downloads the _contents_ of the collection, but not the collection (folder) itself, whereas `iget -r <collection>` does create a local folder `<collection>` and downloads the contents into that. Maybe we should replicate that behaviour.

Thanks for finding it. Fixed!

> 
>     * `irods_put()` seems to overwrite by default, maybe we should add a `overwrite`-flag there as well.

Also a very good point. Fixed.

maartenschermer commented on ibridges/irodsconnector/data_operations.py:
```diff
@@ -110,56 +145,252 @@ def irods_put(self, local_path: str, irods_path: str, resc_name: str = ''):
+                os.makedirs(dest)
+            try:
+                self._obj_get(IrodsPath(self.session, subcoll_path, obj_name),
+                              local_path, overwrite, options)
```
One last thing: I think `local_path` should be `dest` here. Currently, the subcollection's files end up in the root.

chStaiger replied:
Thanks!!!

qubixes commented on ibridges/irodsconnector/data_operations.py:
```diff
@@ -110,56 +145,252 @@ def irods_put(self, local_path: str, irods_path: str, resc_name: str = ''):
+
+        return objs
+
+    def remove(self, irods_path: Union[IrodsPath, str]):
```
Maybe part of IrodsPath as well?

chStaiger replied:
Yes, I already created a copy there. Once the IrodsPath is fixed, we can delete it here.

qubixes reviewed (changes requested):
I don't really have any proper insight, but it looks good to me. I guess you're in the process of rebasing/fixing the CI issues?

chStaiger commented on ibridges/irodsconnector/data_operations.py:
```diff
@@ -1,36 +1,35 @@
         self.session = session
 
-    def get_dataobject(self, path: str) -> irods.data_object.iRODSDataObject:
+    def get_dataobject(self, path: Union[str, IrodsPath]) -> irods.data_object.iRODSDataObject:
```
For now, I would leave it in here. In the python API this function belongs to the data_objects module and it returns a python object of class data_object, while I think it would also be logical to let the IrodsPath only return IrodsPaths, to have a distinct difference between string operations with occasional validation with the Irods server and real operations on the data objects and collections.
For now I would leave it like this and move this discussion to the discussion about our class structure, see below.

chStaiger commented on ibridges/irodsconnector/data_operations.py:
```diff
@@ -85,23 +94,49 @@ def get_collection(self, path: IrodsPath) -> irods.collection.iRODSCollection:
         """
-        if self.collection_exists(path):
-            return self.session.irods_session.collections.get(path)
+        path = IrodsPath(self.session, path)
```
I do not think so. The function takes the whole IrodsPath as a string and checks if the path in its entirety exists on the iRODS server. If so it returns the collection, if not it will raise a CollectionNotFound exception

chStaiger commented on ibridges/irodsconnector/data_operations.py:
```diff
@@ -85,23 +94,49 @@ def get_collection(self, path: IrodsPath) -> irods.collection.iRODSCollection:
+            raise ValueError("local_path must be a file.")
+
+        # Check if irods object already exists
+        obj_exists = False
```
Good point!

maartenschermer commented on ibridges/irodsconnector/session.py:
```diff
@@ -99,9 +99,9 @@ def connect(self):
             #                        host=host)
             raise NotImplementedError
-
+        print(self._password)
```
Left over debug statement, I guess.

chStaiger commented on ibridges/irodsconnector/data_operations.py:
```diff
@@ -110,56 +145,252 @@ def irods_put(self, local_path: str, irods_path: str, resc_name: str = ''):
+        Parameters
+        ----------
+        local_path : Path
+            Absolute path to the directory to upload
```
At the moment just to be sure and to decrease the path fumbling. But we can a extend it with relative paths or automatic extension with home in ibridges.

maartenschermer commented:
I'm testing logging in with the cached password, which works fine if the password has been cached. But if the `.irodsA` file doesn't exist, it gives an error on this line:
https://github.com/UtrechtUniversity/iBridges/blob/07f667ea11a1a18a27a34fe58f3dcc3c12653b46/ibridges/irodsconnector/session.py#L64
because `self._irods_session` doesn't exist (hasn't been defined). If you catch that error, the error correctly falls through to `irods.session.NonAnonymousLoginWithoutPassword: Authentication file not found at '/home/.../.irods/.irodsA'`

qubixes reviewed (approved):
I don't really have anything to add at the moment, except that in the tutorial there is an obj_perm object that probably was renamed to "perm".
